types/index.d.ts

```diff
@@ -0,0 +1,151 @@
+import * as stream from 'stream';
+import { EventEmitter } from 'events'
+
+
+export type Callback<V, Err = any> = (err: Err, value: V) => void;
+
+export type CodecTransformer = (buffer: Buffer, callback: () => void) => Buffer; // TODO
+
+export interface CodecOptions {
+	deflate: CodecTransformer;
+	snappy: CodecTransformer;
+}
+
+export interface Decoder {
+	on(type: 'metadata', callback: (type: Type.Type) => void): this;
+	on(type: 'data', callback: (value: object) => void): this;
+}
+
+export interface Encoder {
+	// TODO
+}
+
+export interface ReaderOptions {
+	// TODO
+}
+
+interface AssembleOptions {
+	importHook: (filePath: string, type: 'idl', callback: Callback<object>) => void;
+}
+
+export function assembleProtocol(filePath: string, opts: AssembleOptions, callback: Callback<object>): void;
+export function assembleProtocol(filePath: string, callback: Callback<object>): void;
+export function createFileDecoder(fileName: string, codecs?: CodecOptions): Decoder;
+export function createFileEncoder(filePath: string, schema: any, options?: any): Encoder;
+export function discoverProtocol(transport: Service.Transport, options: any, callback: Callback<any>): void;
+export function discoverProtocol(transport: Service.Transport, callback: Callback<any>): void;
+export function extractFileHeader(filePath: string, options?: any): void;
+export function parse(schemaOrProtocolIdl: string, options?: any): Service.Protocol | Type.Type; // TODO
+export function readProtocol(protocolIdl: string, options?: ReaderOptions): Service.Protocol;
+export function readSchema(schemaIdl: string, options?: ReaderOptions): Type.Schema;
+// TODO streams
+// TODO types
+
+export namespace Type {
+	type Schema = string;
```

Comment:
Schemas can also be JavaScript objects (for example `{type: 'map', values: 'int'}`).

Reply:
The Js object could also be completely typechecked. Have a look at https://github.com/joewood/avro-typescript/blob/master/src/model.ts

Maybe we could extract a common Ts definition for Avro schemas for use in booth projects.